### PR TITLE
Feature/lazy 68 add trivy scan

### PR DESCRIPTION
**Pull request by naveen-variant**
# Description

This feature adds the ability to scan the component image built as part of the action and look for vulnerabilities . Any vulnerabilities found will fail the run till it get fixed, This is to keep our images safe from security related issues.

Some additional features : 
1 . There might be some low priority or low risk vulnerabilities which we want to skip or can wait for some time to fix  (Note:  Only with devops team approval} .It can be done by creating .trivyignore file and add vulnerability id in the file and upload it in the s3 bucket with the repo name . Action has ability to check from s3 bucket to look for specific repo and pull in trivy ignore file.
2. This action also has  root trivy ignore in case we want any vulnerability need to be skipped or park aside for sometime  across all images.

Fixes [#68](https://usxtech.atlassian.net/browse/LAZY-68)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Unit Tests
- [x] Sanity Testing

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,8 +1,8 @@
-# Actions - Dotnet
+# Lazy Action - Dotnet
```
Comment (vibindaniel) [inline README.md:1]: Rever this to Actions

File: README.md [hunk 2]
```diff
@@ -1,8 +1,8 @@
...
 
 Setting up continuous integration
 
-- [Actions - Dotnet](#actions---dotnet)
+- [Lazy Action - Dotnet](#lazy-action---dotnet)
```
Comment (vibindaniel) [inline README.md:5]: Same here

File: entrypoint.sh [hunk 3]
```diff
@@ -30,6 +30,10 @@ if [ "$INPUT_CONTAINER_PUSH_ENABLED" = 'true' ]; then
   echo "End: Publish Image to ECR"
 fi
 
+echo "Start: Trivy Scan"
```
Comment (vibindaniel) [inline entrypoint.sh:33]: This should be run before docker push is done. It makes **0** sense to do it after docker push.

File: scripts/trivy_scan.sh [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+aws configure set aws_access_key_id "$(echo "$credentials" | jq -r '.Credentials.AccessKeyId')" --profile ops
+aws configure set aws_secret_access_key "$(echo "$credentials" | jq -r '.Credentials.SecretAccessKey')" --profile ops
+aws configure set aws_session_token "$(echo "$credentials" | jq -r '.Credentials.SessionToken')" --profile ops
+
+echo "Print repo name: $GITHUB_REPOSITORY"
+echo "Download root trivy file from s3" 
+eval "aws --profile ops s3 cp s3://${S3_BUCKET_NAME}/.trivyignore ."
+
+PATH_TO_FOLDER=$GITHUB_REPOSITORY
+totalFoundObjects="$(aws s3 --profile ops ls s3://"${S3_BUCKET_NAME}"/"${PATH_TO_FOLDER}"/.trivyignore --recursive --summarize | grep "Total Objects:" | sed 's/[^0-9]*//g')"
```
Comment (vibindaniel) [inline scripts/trivy_scan.sh:22]: Refactor this and use `aws cp`. If `aws cp` fails then you know that there is no file like that in the s3 bucket.

**Review by vibindaniel — CHANGES_REQUESTED** [review]

**Comment by vibindaniel** [comment]
@naveen-variant Please set up gpg signature verfication as mentioned in #lazy-devs

File: README.md (hunk 2 above)
Comment (naveen-variant) [reply README.md]: Updated.

File: entrypoint.sh (hunk 3 above)
Comment (naveen-variant) [reply entrypoint.sh]: Refactored to run trivy push and then docker push.

File: scripts/trivy_scan.sh (hunk 4 above)
Comment (naveen-variant) [reply scripts/trivy_scan.sh]: Any issue with aws s3 listing . If we have to go with fail command approach , again we have to unset error (like set +e out properties  and  enable it back . If there is any pressing reason , i also have changes ready for cp command fail approach .

File: README.md (hunk 1 above)
Comment (naveen-variant) [reply README.md]: Updated

File: scripts/trivy_scan.sh (hunk 4 above)
Comment (naveen-variant) [reply scripts/trivy_scan.sh]: I have changed to aws cp though.

**Comment by naveen-variant** [comment]
@salyateem13 This PR can be tested by create a commit in demo-app (https://github.com/variant-inc/demo-app) .Need to verify logs for trivy scan and see it is downloading .trivyignore file from s3 and using that file for trivy scan the image generated with that run.

**Comment by naveen-variant** [comment]
> @naveen-variant Please set up gpg signature verfication as mentioned in #lazy-devs

Added .All the comments from now will have verified tag.

**Comment by salyateem13** [comment]
> @salyateem13 This PR can be tested by create a commit in demo-app (https://github.com/variant-inc/demo-app) .Need to verify logs for trivy scan and see it is downloading .trivyignore file from s3 and using that file for trivy scan the image generated with that run.

@naveen-variant Looks like the addition of trivy to the action works fine, but fails later in the action on dotnet nuget push command. https://github.com/variant-inc/demo-app/runs/2278292379?check_suite_focus=true#step:5:1182. What is the issue here. If its non issue I can approve.

**Comment by naveen-variant** [comment]
> > @salyateem13 This PR can be tested by create a commit in demo-app (https://github.com/variant-inc/demo-app) .Need to verify logs for trivy scan and see it is downloading .trivyignore file from s3 and using that file for trivy scan the image generated with that run.
> 
> @naveen-variant Looks like the addition of trivy to the action works fine, but fails later in the action on dotnet nuget push command. https://github.com/variant-inc/demo-app/runs/2278292379?check_suite_focus=true#step:5:1182. What is the issue here. If its non issue I can approve.

@salyateem13 Yes nuget failure can be ignored.

**Review by vibindaniel — APPROVED** [review]

File: action.yml [hunk 5]
```diff
@@ -37,6 +37,6 @@ inputs:
 
 runs:
   using: "docker"
-  image: docker://public.ecr.aws/p4v7w0a5/lazy/dotnet-action:v0.1.5
+  image: docker://public.ecr.aws/p4v7w0a5/lazy/dotnet-action:v0.1.6
```
Comment (vibindaniel) [inline action.yml:40]: Why would you go from v1 to v0.1.6? Shouldn't it be v1.0.1 at least?

**Review by vibindaniel — CHANGES_REQUESTED** [review]

File: action.yml (hunk 5 above)
Comment (naveen-variant) [reply action.yml:40]: If we are considering this as major release then we can tag it as  v1.0.1 .

Comment (vibindaniel) [reply action.yml:40]: The next major release will be v2.

Comment (naveen-variant) [reply action.yml:40]: ok will create release as v1.0.1 . Are we still going to keep the tag as v1 .

Comment (vibindaniel) [reply action.yml]: Yes, that is correct.

Comment (naveen-variant) [reply action.yml]: ok can you please approve the PR . I will go ahead and merge .

**Review by vibindaniel — APPROVED** [review]

**Pull request merged**